File: sdk/go/query/query.go
// Package query 提供 torchwood 文档查询的 typed 构造器与 DSL 语法糖
// （C7 单 AST：服务端只收 shared.v1.Query；DSL 串在客户端解析为 AST 后发送）。
//
// 构造器产出 *sharedv1.Query，供 sdk/go 的 client/server 两面与直接调用
// gRPC/HTTP 网关的用户使用。DSL 语法糖 FromDSL 的文法与仓库根模块的
// pkg/query 同源（Appwrite 风格）；本包自带解析器以保持 SDK 模块的依赖
// 面最小（仅 genproto + stdlib），文法演进需两侧同步（sdk/go/query 与
// pkg/query 的 golden 用例锁对齐）。
package query

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"

	sharedv1 "github.com/torchwooddev/torchwood/genproto/shared/v1"
)

// ---------------------------------------------------------------------------
// Filter 叶子构造器（Comparison 形态）
// ---------------------------------------------------------------------------

// comparison 包一个 Comparison。
func comparison(attr string, values []string) *sharedv1.Comparison {
	return &sharedv1.Comparison{Attribute: attr, Values: values}
}

// Eq 等值；多值时语义为 IN（与服务端编译一致）。
func Eq(attr string, values ...string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_Eq{Eq: comparison(attr, values)}}
}

// Ne 不等；多值时语义为 NOT IN。
func Ne(attr string, values ...string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_Ne{Ne: comparison(attr, values)}}
}

func Lt(attr, value string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_Lt{Lt: comparison(attr, []string{value})}}
}

func Lte(attr, value string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_Lte{Lte: comparison(attr, []string{value})}}
}

func Gt(attr, value string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_Gt{Gt: comparison(attr, []string{value})}}
}

func Gte(attr, value string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_Gte{Gte: comparison(attr, []string{value})}}
}

// In 集合成员（值列表）。
func In(attr string, values ...string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_In{In: comparison(attr, values)}}
}

func Contains(attr, value string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_Contains{Contains: comparison(attr, []string{value})}}
}

func NotContains(attr, value string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_NotContains{NotContains: comparison(attr, []string{value})}}
}

func StartsWith(attr, value string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_StartsWith{StartsWith: comparison(attr, []string{value})}}
}

func NotStartsWith(attr, value string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_NotStartsWith{NotStartsWith: comparison(attr, []string{value})}}
}

func EndsWith(attr, value string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_EndsWith{EndsWith: comparison(attr, []string{value})}}
}

func NotEndsWith(attr, value string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_NotEndsWith{NotEndsWith: comparison(attr, []string{value})}}
}

// Search 全文检索（目标属性须有 fulltext 索引）。
func Search(attr, value string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_Search{Search: comparison(attr, []string{value})}}
}

func NotSearch(attr, value string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_NotSearch{NotSearch: comparison(attr, []string{value})}}
}

// Between 闭区间 [min, max]。
func Between(attr, min, max string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_Between{Between: comparison(attr, []string{min, max})}}
}

func NotBetween(attr, min, max string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_NotBetween{NotBetween: comparison(attr, []string{min, max})}}
}

// IsNull / IsNotNull 空值判定（无值算子）。
func IsNull(attr string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_IsNull{IsNull: comparison(attr, nil)}}
}

func IsNotNull(attr string) *sharedv1.Filter {
	return &sharedv1.Filter{Expr: &sharedv1.Filter_IsNotNull{IsNotNull: comparison(attr, nil)}}
}

// And / Or 组合布尔节点（嵌套深度 ≤8）；nil 子节点被跳过，单个子节点坍缩。
func And(children ...*sharedv1.Filter) *sharedv1.Filter {
	kept := keepNonNil(children)
	switch len(kept) {
	case 0:
		return nil
	case 1:
		return kept[0]
	}
	return &sharedv1.Filter{Expr: &sharedv1.Filter_And{And: &sharedv1.FilterList{Filters: kept}}}
}

func Or(children ...*sharedv1.Filter) *sharedv1.Filter {
	kept := keepNonNil(children)
	switch len(kept) {
	case 0:
		return nil
	case 1:
		return kept[0]
	}
	return &sharedv1.Filter{Expr: &sharedv1.Filter_Or{Or: &sharedv1.FilterList{Filters: kept}}}
}

func keepNonNil(children []*sharedv1.Filter) []*sharedv1.Filter {
	kept := make([]*sharedv1.Filter, 0, len(children))
	for _, c := range children {
		if c != nil {
			kept = append(kept, c)
		}
	}
	return kept
}

// ---------------------------------------------------------------------------
// Query 构造器（链式）
// ---------------------------------------------------------------------------

// Builder 链式构造 shared.v1.Query。
type Builder struct {
	q sharedv1.Query
}

// New 创建空 Builder。
func New() *Builder { return &Builder{} }

// Filter 设置过滤树（多次调用后者覆盖）。
func (b *Builder) Filter(f *sharedv1.Filter) *Builder {
	b.q.Filter = f
	return b
}

// OrderAsc 追加升序排序键（服务端强制追加 $id tiebreaker）。
func (b *Builder) OrderAsc(attr string) *Builder {
	b.q.Orders = append(b.q.Orders, &sharedv1.Order{Attribute: attr})
	return b
}

// OrderDesc 追加降序排序键。
func (b *Builder) OrderDesc(attr string) *Builder {
	b.q.Orders = append(b.q.Orders, &sharedv1.Order{Attribute: attr, Desc: true})
	return b
}

// Select 投影：只返回这些字段。
func (b *Builder) Select(fields ...string) *Builder {
	b.q.Select = append(b.q.Select, fields...)
	return b
}

// PageSize 页大小（1..200；0 = 服务端默认）。
func (b *Builder) PageSize(n int32) *Builder {
	b.q.PageSize = n
	return b
}

// PageToken 续页 token（服务端返回的 ka:/kb: keyset token）。
func (b *Builder) PageToken(tok string) *Builder {
	b.q.PageToken = tok
	return b
}

// Build 产出 *sharedv1.Query（proto 消息含锁不可值拷贝，逐字段组装）。
func (b *Builder) Build() *sharedv1.Query {
	out := &sharedv1.Query{
		PageSize:  b.q.GetPageSize(),
		PageToken: b.q.GetPageToken(),
	}
	out.Filter = b.q.GetFilter()
	out.Orders = append(out.Orders, b.q.GetOrders()...)
	out.Select = append(out.Select, b.q.GetSelect()...)
	return out
}

// ---------------------------------------------------------------------------
// DSL 语法糖：Appwrite 风格串 → *sharedv1.Query（客户端解析，服务端零消费）
// ---------------------------------------------------------------------------

var dslRe = regexp.MustCompile(`^(\w+)\((.*)\)$`)

// FromDSL 把 Appwrite 风格 DSL 串解析为 typed Query（隐式 AND 合并）。
// 支持算子：equal/notEqual/lessThan/lessThanEqual/greaterThan/
// greaterThanEqual/in/contains/notContains/startsWith/notStartsWith/
// endsWith/notEndsWith/search/notSearch/between/notBetween/isNull/
// isNotNull/orderAsc/orderDesc/limit/select/cursorAfter/cursorBefore。
// offset() 不支持（文档面 keyset-only）；cursorAfter/Before 映射为
// ka:/kb: keyset page token。
func FromDSL(parts ...string) (*sharedv1.Query, error) {
	q := &sharedv1.Query{}
	var leaves []*sharedv1.Filter
	for _, raw := range parts {
		raw = strings.TrimSpace(raw)
		if raw == "" {
			continue
		}
		m := dslRe.FindStringSubmatch(raw)
		if m == nil {
			return nil, fmt.Errorf("invalid query format: %s", raw)
		}
		op, argStr := m[1], m[2]
		args, err := splitArgs(argStr)
		if err != nil {
			return nil, fmt.Errorf("failed to parse query %q: %w", raw, err)
		}
		switch op {
		case "equal", "notEqual", "lessThan", "lessThanEqual", "greaterThan", "greaterThanEqual",
			"contains", "notContains", "startsWith", "notStartsWith", "endsWith", "notEndsWith",
			"search", "notSearch":
			if len(args) < 2 {
				return nil, fmt.Errorf("%s requires at least 2 args", op)
			}
			attr, err := unquote(args[0])
			if err != nil {
				return nil, err
			}
			var values []string
			if strings.HasPrefix(args[1], "[") {
				values, err = parseArray(args[1])
				if err != nil {
					return nil, err
				}
			} else {
				v, err := unquoteOrLiteral(args[1])
				if err != nil {
					return nil, err
				}
				values = []string{v}
			}
			leaf, err := dslLeaf(op, attr, values)
			if err != nil {
				return nil, err
			}
			leaves = append(leaves, leaf)
		case "in":
			if len(args) != 2 {
				return nil, fmt.Errorf("in requires 2 args")
			}
			attr, err := unquote(args[0])
			if err != nil {
				return nil, err
			}
			if !strings.HasPrefix(args[1], "[") {
				return nil, fmt.Errorf("in requires an array of values, e.g. in(%q, [\"a\",\"b\"])", attr)
			}
			values, err := parseArray(args[1])
			if err != nil {
				return nil, err
			}
			if len(values) == 0 {
				return nil, fmt.Errorf("in requires at least 1 value")
			}
			leaves = append(leaves, In(attr, values...))
		case "between", "notBetween":
			if len(args) != 3 {
				return nil, fmt.Errorf("%s requires 3 args", op)
			}
			attr, err := unquote(args[0])
			if err != nil {
				return nil, err
			}
			min, err := unquoteOrLiteral(args[1])
			if err != nil {
				return nil, err
			}
			max, err := unquoteOrLiteral(args[2])
			if err != nil {
				return nil, err
			}
			leaves = append(leaves, dslBetween(op, attr, min, max))
		case "isNull", "isNotNull":
			if len(args) != 1 {
				return nil, fmt.Errorf("%s requires 1 arg", op)
			}
			attr, err := unquote(args[0])
			if err != nil {
				return nil, err
			}
			if op == "isNull" {
				leaves = append(leaves, IsNull(attr))
			} else {
				leaves = append(leaves, IsNotNull(attr))
			}
		case "orderAsc", "orderDesc":
			if len(args) != 1 {
				return nil, fmt.Errorf("%s requires 1 arg", op)
			}
			attr, err := unquote(args[0])
			if err != nil {
				return nil, err
			}
			q.Orders = append(q.Orders, &sharedv1.Order{Attribute: attr, Desc: op == "orderDesc"})
		case "limit":
			if len(args) != 1 {
				return nil, fmt.Errorf("limit requires 1 arg")
			}
			n, err := strconv.Atoi(args[0])
			if err != nil || n < 0 {
				return nil, fmt.Errorf("limit must be a non-negative integer")
			}
			if n > 0 {
				q.PageSize = int32(n)
			}
		case "select":
			if len(args) != 1 {
				return nil, fmt.Errorf("select requires 1 arg")
			}
			fields, err := parseArray(args[0])
			if err != nil {
				return nil, err
			}
			q.Select = append(q.Select, fields...)
		case "cursorAfter", "cursorBefore":
			if len(args) != 1 {
				return nil, fmt.Errorf("%s requires 1 arg", op)
			}
			id, err := unquote(args[0])
			if err != nil {
				return nil, err
			}
			if op == "cursorAfter" {
				q.PageToken = "ka:" + id
			} else {
				q.PageToken = "kb:" + id
			}
		case "offset":
			return nil, fmt.Errorf("offset() is not supported; documents use cursor (keyset) pagination")
		default:
			return nil, fmt.Errorf("unsupported query operator: %s", op)
		}
	}
	if len(leaves) == 1 {
		q.Filter = leaves[0]
	} else if len(leaves) > 1 {
		q.Filter = And(leaves...)
	}
	return q, nil
}

func dslLeaf(op, attr string, values []string) (*sharedv1.Filter, error) {
	if len(values) == 0 {
		return nil, fmt.Errorf("%s requires at least 1 value", op)
	}
	switch op {
	case "equal":
		return Eq(attr, values...), nil
	case "notEqual":
		return Ne(attr, values...), nil
	case "lessThan":
		return Lt(attr, values[0]), nil
	case "lessThanEqual":
		return Lte(attr, values[0]), nil
	case "greaterThan":
		return Gt(attr, values[0]), nil
	case "greaterThanEqual":
		return Gte(attr, values[0]), nil
	case "contains":
		return Contains(attr, values[0]), nil
	case "notContains":
		return NotContains(attr, values[0]), nil
	case "startsWith":
		return StartsWith(attr, values[0]), nil
	case "notStartsWith":
		return NotStartsWith(attr, values[0]), nil
	case "endsWith":
		return EndsWith(attr, values[0]), nil
	case "notEndsWith":
		return NotEndsWith(attr, values[0]), nil
	case "search":
		return Search(attr, values[0]), nil
	default:
		return NotSearch(attr, values[0]), nil
	}
}

func dslBetween(op, attr, min, max string) *sharedv1.Filter {
	if op == "between" {
		return Between(attr, min, max)
	}
	return NotBetween(attr, min, max)
}

// splitArgs 按顶层逗号分割（尊重引号与括号嵌套）。
func splitArgs(s string) ([]string, error) {
	var args []string
	var sb strings.Builder
	depth := 0
	inQuote := false
	var escape bool
	for _, r := range s {
		if escape {
			sb.WriteRune(r)
			escape = false
			continue
		}
		if r == '\\' {
			escape = true
			sb.WriteRune(r)
			continue
		}
		if r == '"' {
			inQuote = !inQuote
			sb.WriteRune(r)
			continue
		}
		if inQuote {
			sb.WriteRune(r)
			continue
		}
		switch r {
		case '[', '(':
			depth++
			sb.WriteRune(r)
		case ']', ')':
			depth--
			sb.WriteRune(r)
		case ',':
			if depth == 0 {
				args = append(args, strings.TrimSpace(sb.String()))
				sb.Reset()
			} else {
				sb.WriteRune(r)
			}
		default:
			sb.WriteRune(r)
		}
	}
	if inQuote || depth != 0 {
		return nil, fmt.Errorf("unbalanced arguments: %s", s)
	}
	if sb.Len() > 0 {
		args = append(args, strings.TrimSpace(sb.String()))
	}
	return args, nil
}

func unquote(s string) (string, error) {
	s = strings.TrimSpace(s)
	if len(s) < 2 || s[0] != '"' || s[len(s)-1] != '"' {
		return "", fmt.Errorf("expected quoted string, got %s", s)
	}
	return unescape(s[1 : len(s)-1]), nil
}

func unquoteOrLiteral(s string) (string, error) {
	s = strings.TrimSpace(s)
	if len(s) >= 2 && s[0] == '"' && s[len(s)-1] == '"' {
		return unescape(s[1 : len(s)-1]), nil
	}
	return s, nil
}

func unescape(s string) string {
	var b strings.Builder
	for i := 0; i < len(s); i++ {
		if s[i] == '\\' && i+1 < len(s) {
			b.WriteByte(s[i+1])
			i++
			continue
		}
		b.WriteByte(s[i])
	}
	return b.String()
}

func parseArray(s string) ([]string, error) {
	s = strings.TrimSpace(s)
	if len(s) < 2 || s[0] != '[' || s[len(s)-1] != ']' {
		return nil, fmt.Errorf("expected array, got %s", s)
	}
	items, err := splitArgs(s[1 : len(s)-1])
	if err != nil {
		return nil, err
	}
	for i := range items {
		v, err := unquoteOrLiteral(items[i])
		if err != nil {
			return nil, err
		}
		items[i] = v
	}
	return items, nil
}

File: sdk/typescript/src/query.ts
// 文档查询 typed AST（C7 单 AST）：服务端只收 shared.v1.Query；
// 本模块是 TS 侧的类型与构造器。字段名与 protojson 对齐
// （filter/orders/select/pageSize/pageToken；oneof 分支 eq/ne/…/notBetween/
// isNull/and/or）。

/** 比较叶：attribute + values（isNull/isNotNull 无 values；between 恰 2 个）。 */
export interface QueryComparison {
  attribute: string;
  values?: string[];
}

/** 布尔组合：and/or 嵌套（深度 ≤8）。 */
export interface FilterList {
  filters: FilterNode[];
}

export type FilterNode =
  | { eq: QueryComparison }
  | { ne: QueryComparison }
  | { lt: QueryComparison }
  | { lte: QueryComparison }
  | { gt: QueryComparison }
  | { gte: QueryComparison }
  | { in: QueryComparison }
  | { contains: QueryComparison }
  | { notContains: QueryComparison }
  | { startsWith: QueryComparison }
  | { notStartsWith: QueryComparison }
  | { endsWith: QueryComparison }
  | { notEndsWith: QueryComparison }
  | { search: QueryComparison }
  | { notSearch: QueryComparison }
  | { between: QueryComparison }
  | { notBetween: QueryComparison }
  | { isNull: QueryComparison }
  | { isNotNull: QueryComparison }
  | { containsAny: QueryComparison }
  | { containsAll: QueryComparison }
  | { and: FilterList }
  | { or: FilterList };

/** 单个排序键（服务端强制追加 $id tiebreaker）。 */
export interface QueryOrder {
  attribute: string;
  desc?: boolean;
}

/** KNN 距离度量（与 hnsw 索引的 distance_metric 同枚举域）。 */
export type DistanceMetric = "COSINE" | "L2" | "INNER_PRODUCT";

/**
 * 向量近邻查询（会话 #10 §10.5 P0；B2 多页）：非 filter 树节点——距离承载
 * 排序，pageSize 即 k（top-k 可见近邻）。attribute 须为声明的 vector 属性且
 * 存在 metric 匹配的 hnsw 索引；与 orders 互斥（服务端拒绝）。多页翻页：
 * 同 Query 以 pageToken 携带服务端发放的 kvc: 距离游标（跨页不重不漏）。
 * DSL 字符串不支持 vector_search（向量不该手写，typed builder only）。
 */
export interface VectorSearch {
  attribute: string;
  values: number[];
  metric?: DistanceMetric;
  maxDistance?: number;
  /**
   * HNSW 搜索广度（B7）：本次查询的 hnsw.ef_search，合法域 [1,500]
   * （≤0 / >500 服务端 InvalidArgument）。缺省不发送——服务端维持 pgvector
   * 缺省 40。近重复簇边界召回不足时调大（代价：延迟随 ef 增长）。
   */
  efSearch?: number;
}

/** shared.v1.Query 的 JSON 形态（POST documents:list 的 body 即此对象）。 */
export interface QueryAst {
  filter?: FilterNode;
  orders?: QueryOrder[];
  select?: string[];
  pageSize?: number;
  pageToken?: string;
  vectorSearch?: VectorSearch;
}

function cmp(attribute: string, values: string[]): QueryComparison {
  return { attribute, values };
}

/** 等值；多值时语义为 IN（与服务端编译一致）。 */
export function eq(attribute: string, ...values: string[]): FilterNode {
  return { eq: cmp(attribute, values) };
}

/** 不等；多值时语义为 NOT IN。 */
export function ne(attribute: string, ...values: string[]): FilterNode {
  return { ne: cmp(attribute, values) };
}

export function lt(attribute: string, value: string): FilterNode {
  return { lt: cmp(attribute, [value]) };
}

export function lte(attribute: string, value: string): FilterNode {
  return { lte: cmp(attribute, [value]) };
}

export function gt(attribute: string, value: string): FilterNode {
  return { gt: cmp(attribute, [value]) };
}

export function gte(attribute: string, value: string): FilterNode {
  return { gte: cmp(attribute, [value]) };
}

export function in_(attribute: string, values: string[]): FilterNode {
  return { in: cmp(attribute, values) };
}

export function contains(attribute: string, value: string): FilterNode {
  return { contains: cmp(attribute, [value]) };
}

export function notContains(attribute: string, value: string): FilterNode {
  return { notContains: cmp(attribute, [value]) };
}

export function startsWith(attribute: string, value: string): FilterNode {
  return { startsWith: cmp(attribute, [value]) };
}

export function notStartsWith(attribute: string, value: string): FilterNode {
  return { notStartsWith: cmp(attribute, [value]) };
}

export function endsWith(attribute: string, value: string): FilterNode {
  return { endsWith: cmp(attribute, [value]) };
}

export function notEndsWith(attribute: string, value: string): FilterNode {
  return { notEndsWith: cmp(attribute, [value]) };
}

/** 全文检索（目标属性须有 fulltext 索引）。 */
export function search(attribute: string, value: string): FilterNode {
  return { search: cmp(attribute, [value]) };
}

export function notSearch(attribute: string, value: string): FilterNode {
  return { notSearch: cmp(attribute, [value]) };
}

/** 闭区间 [min, max]。 */
export function between(attribute: string, min: string, max: string): FilterNode {
  return { between: cmp(attribute, [min, max]) };
}

export function notBetween(attribute: string, min: string, max: string): FilterNode {
  return { notBetween: cmp(attribute, [min, max]) };
}

export function isNull(attribute: string): FilterNode {
  return { isNull: { attribute } };
}

export function isNotNull(attribute: string): FilterNode {
  return { isNotNull: { attribute } };
}

/**
 * 数组算子（§10.5 P0）：仅 array=true 属性可用（服务端按 catalog attrs
 * 白名单校验）。containsAny = 交集非空（PG &&）；containsAll = 子集（PG @>）。
 */
export function containsAny(attribute: string, values: string[]): FilterNode {
  return { containsAny: cmp(attribute, values) };
}

export function containsAll(attribute: string, values: string[]): FilterNode {
  return { containsAll: cmp(attribute, values) };
}

/** and/or 组合（nil 节点由调用方自行过滤）。 */
export function and(...filters: FilterNode[]): FilterNode {
  return { and: { filters } };
}

export function or(...filters: FilterNode[]): FilterNode {
  return { or: { filters } };
}

export function orderAsc(attribute: string): QueryOrder {
  return { attribute };
}

export function orderDesc(attribute: string): QueryOrder {
  return { attribute, desc: true };
}

/**
 * KNN 构造器（metric 缺省 COSINE）。示例：
 *   vectorSearch("emb", [0.1, 0.2, 0.3]).metric("L2").maxDistance(0.5)
 * values 须与目标 vector 属性的声明维度等长（服务端按 catalog 校验）。
 */
export function vectorSearch(attribute: string, values: number[]) {
  const v: VectorSearch = { attribute, values };
  return {
    /** 距离度量（须与目标列 hnsw 索引的 distance_metric 匹配）。 */
    metric(m: DistanceMetric) {
      v.metric = m;
      return this;
    },
    /** 距离阈值：仅保留 top-k 中距离 <= max 的行。 */
    maxDistance(max: number) {
      v.maxDistance = max;
      return this;
    },
    /** HNSW 搜索广度（B7）：hnsw.ef_search，合法域 [1,500]；缺省不发送（服务端用 pgvector 缺省 40）。 */
    efSearch(n: number) {
      v.efSearch = n;
      return this;
    },
    build(): VectorSearch {
      return v;
    },
  };
}
